In [14]:
import numpy as np
import layers
import activations
import nn
import losses
import optimizers

In [17]:
from tensorflow.keras.datasets import mnist
import numpy as np

#load and preprocess the MNIST dataset
(X_train, y_train), (X_test, y_test) = mnist.load_data()
X_train = X_train.reshape(-1, 784) / 255.0
X_test = X_test.reshape(-1, 784) / 255.0

#convert labels to one-hot encoded vectors
def one_hot_encode(labels, num_classes):
    num_samples = labels.shape[0]
    encoded_labels = np.zeros((num_samples, num_classes))
    for i in range(num_samples):
        encoded_labels[i, labels[i]] = 1
    return encoded_labels

num_classes = 10
y_train_one_hot = one_hot_encode(y_train, num_classes)
y_test_one_hot = one_hot_encode(y_test, num_classes)


#create the model
input_size = 784
num_units = 128
model = nn.Sequential()
model.add(layers.InputLayer(input_shape=input_size))
model.add(layers.Dense(input_size, num_units, activation=activations.Sigmoid, optimizer=optimizers.RMSprop))
model.add(layers.Dense(num_units, num_classes, activation=activations.Softmax, optimizer=optimizers.RMSprop))

#train the model
epochs = 25
learning_rate = 0.001
batch_size = 128
model.train(X_train.T, y_train_one_hot.T, epochs, learning_rate, batch_size, loss=losses.CrossEntropy)

#make predictions
predictions = model.predict(X_test.T)

#convert predictions to class labels
predicted_labels = np.argmax(predictions, axis=0)

#calculate accuracy
accuracy = np.mean(predicted_labels == y_test)
print("Accuracy:", accuracy)

Epoch 0, Loss: 0.30356490386062646
Epoch 1, Loss: 0.23504362040619364
Epoch 2, Loss: 0.19612274951488484
Epoch 3, Loss: 0.16797468227813567
Epoch 4, Loss: 0.14644867613389823
Epoch 5, Loss: 0.1294722537376399
Epoch 6, Loss: 0.11574242537069016
Epoch 7, Loss: 0.10439458473397185
Epoch 8, Loss: 0.09484843609464838
Epoch 9, Loss: 0.08669522138857041
Epoch 10, Loss: 0.07963410219569332
Epoch 11, Loss: 0.07344087206348919
Epoch 12, Loss: 0.06794782666485476
Epoch 13, Loss: 0.06303049363608199
Epoch 14, Loss: 0.05859586367442724
Epoch 15, Loss: 0.05457437659587271
Epoch 16, Loss: 0.05091345085475838
Epoch 17, Loss: 0.04757066223422287
Epoch 18, Loss: 0.044509112978226226
Epoch 19, Loss: 0.04169497060456296
Epoch 20, Loss: 0.03909764585711523
Epoch 21, Loss: 0.03669156082635241
Epoch 22, Loss: 0.03445495469230269
Epoch 23, Loss: 0.0323694031847588
Epoch 24, Loss: 0.03041958855077139
Accuracy: 0.9774


As you can see above, using the Sigmoid activation function, the Cross Entropy loss, and the RMSprop optimizer, we achieve an accuracy of 0.9774